src/assignment-5/CollateralizedVault.sol

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import {IERC20} from "yield-utils-v2/token/IERC20.sol";
+import {AggregatorV3Interface} from "chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
+import {Ownable} from "openzeppelin/contracts/access/Ownable.sol";
+
+import "forge-std/console2.sol";
+
+interface IERC20WithDecimals is IERC20 {
+    function decimals() external view returns (uint8);
+}
+
+/// @title Collateralized Vault
+/// @author davidbrai
+/// @notice A vault that allows to borrow a token against a collateral
+///     If a debt position is not sufficiently collateralized, e.g when the market price
+///     of the collateral goes down, then the user's collateral may be liquidated by the admin.
+///     In case of a liquidation, both the user's debt and collateral are erased.
+/// @dev The Vault uses Chainlink price feeds to determine the value of the collateral compared to the debt
+contract CollateralizedVault is Ownable {
+
+    /******************
+     * Immutables
+     ******************/
+
+    /// @notice ERC20 token which can be borrowed from the vault
+    IERC20WithDecimals public immutable underlying;
+
+    /// @notice ERC20 token which can be used as collateral
+    IERC20WithDecimals public immutable collateral;
+
+    /// @notice Chainlink priceFeed of underlying / collateral
+    ///     e.g. if underlying is DAI and collateral is WETH, priceFeed is for DAI/WETH
+    AggregatorV3Interface public immutable oracle;
+
+    /// @dev Number of decimals for underlying token
+    uint8 public immutable underlyingDecimals;
```

Comment:
Maybe you can find a way to pack this with `underlying` in a struct, and save a SLOAD somewhere?